From 935ee35a710e7d2bb2b23a055fb8e6c2a2959c36 Mon Sep 17 00:00:00 2001
From: Thierry Bultel <thierry.bultel@iot.bzh>
Date: Fri, 15 Feb 2019 14:54:07 +0100
Subject: [PATCH 1/1] other apps than oneself should be able to display apps

Signed-off-by: Thierry Bultel <thierry.bultel@iot.bzh>
---
 src/hs-client.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/hs-client.cpp b/src/hs-client.cpp
index dd91e10..216f97f 100644
--- a/src/hs-client.cpp
+++ b/src/hs-client.cpp
@@ -226,7 +226,7 @@ int HS_Client::showWindow(afb_req_t request)
     hs_add_object_to_json_object_str( push_obj, 4, _application_id, my_id.c_str(), _type, __FUNCTION__);
     const char* param = afb_req_value(request, _parameter);
     if(param) {
-        std::string req_appid = std::move(get_application_id(request));
+        std::string req_appid = std::move(my_id.c_str());
         if(req_appid.empty()) {
             HMI_NOTICE("homescreen-service","can't get application identifier");
             return AFB_REQ_GETAPPLICATIONID_ERROR;
-- 
2.16.4

