--- ./configure.ac	2017-05-24 18:44:21.927602819 +0200
+++ ./configure.ac	2017-05-24 18:44:51.956866826 +0200
@@ -196,8 +196,7 @@
 			AC_MSG_RESULT(yes)
 			LIBMXML="libmxml.so.1.5"
 			DSO="\$(CC)"
-			DSOFLAGS="$DSOFLAGS -Wl,-soname,libmxml.so.1,-rpath,\$(libdir) -shared \$(OPTIM)"
+			DSOFLAGS="$DSOFLAGS -Wl,-soname,libmxml.so.1 -shared \$(OPTIM)"
-                        LDFLAGS="$LDFLAGS -Wl,-rpath,\$(libdir)"
 			;;
 
 		*BSD*)
